### PR TITLE
Add variant support to PropertyAdd plugin

### PR DESCRIPTION
**Pull request by iammart**
Following on from the discussion on the below thread, I have updated the plugin to query on product code(s) not product ID.

This enables product attributes to be stored on order products that have types of `selection` or `bundle`.

https://aimeos.org/help/viewtopic.php?p=15713#p15713

### CODE REVIEW

**Comment by aimeos** [comment]
Using codes only leads to the situation that it now works for variant articles but not for selection products any more. You need to fetch all products by their ID or their code.

**Comment by iammart** [comment]
If I update the existing expression to following:

```php
$expr = [
    $search->combine('||', [
        $search->compare('==', 'product.id', array_unique($productIds)),
        $search->compare('==', 'product.code', array_unique($productCodes)),
    ]),
    $search->getConditions()
];
$search->setConditions($search->combine('&&', $expr));
```

And the `getProductItems` signature to accept an additional property `$productCodes`,  I am now returned two products in the list.

The `addAttributes` filters the product from the array list using the productId stored on the order. Do we need to check here on the product type and switch between code and ID in order to get the correct product reference to grab the properties we are attaching to the order product?

**Comment by aimeos** [comment]
It's expected that it returns two products for ordered variant articles (the article and the selection product) and we want to get properties from both if they exists.

You can shorten the code a bit:
```
$search->add( $search->or( [
    $search->is( 'product.id', '==', array_unique( $productIds ) ),
    $search->is( 'product.code', '==', array_unique( $productCodes ) ),
] );
$manager->search( $search );
```

**Comment by iammart** [comment]
That makes sense now, I was ignoring the possibility the article could still have properties that might need to be stored against the order product. This now introduces more questions :) Which properties should take priority?

For eg. an article and selection will have the same property options. If one of these is a radio group, either yes/no then there is always a value. This would be stored as two json objects, and result in two comma separated values when printed out in the back office under the order.

I'll make the changes for now and we can review.

**Comment by aimeos** [comment]
If both, selection products and variant articles has the same property type assigned, then article properties should have priority - but I wouldn't care much about that scenario yet

File: lib/mshoplib/src/MShop/Plugin/Provider/Order/PropertyAdd.php [hunk 1]
```diff
@@ -121,18 +121,19 @@ public function update( \Aimeos\MW\Observer\Publisher\Iface $order, string $acti
 		if( !is_array( $value ) )
 		{
 			\Aimeos\MW\Common\Base::checkClass( \Aimeos\MShop\Order\Item\Base\Product\Iface::class, $value );
-			return $this->addAttributes( $value, $this->getProductItems( [$value->getProductId()] ), $types );
+			return $this->addAttributes( $value, $this->getProductItems( [$value->getProductId()], [$value->getProductCode()]), $types );
 		}
 
 		$list = [];
```
Comment (aimeos) [inline lib/mshoplib/src/MShop/Plugin/Provider/Order/PropertyAdd.php:127]: Would be better to use `$ids` and `$codes` instead of a two-dimensional array

File: lib/mshoplib/src/MShop/Plugin/Provider/Order/PropertyAdd.php [hunk 2]
```diff
@@ -121,18 +121,19 @@ public function update( \Aimeos\MW\Observer\Publisher\Iface $order, string $acti
...
 
 		foreach( $value as $orderProduct )
 		{
-			\Aimeos\MW\Common\Base::checkClass( \Aimeos\MShop\Order\Item\Base\Product\Iface::class, $orderProduct );
-			$list[] = $orderProduct->getProductId();
-		}
+			\Aimeos\MW\Common\Base::checkClass(\Aimeos\MShop\Order\Item\Base\Product\Iface::class, $orderProduct);
+			$list['code'][] = $orderProduct->getProductCode();
+			$list['id'][] = $orderProduct->getProductId();
+        	}
```
Comment (aimeos) [inline lib/mshoplib/src/MShop/Plugin/Provider/Order/PropertyAdd.php:134]: You can shorten this to:
```
$ids = map( $value )->getProductId();
$codes = map( $value )->getProductCode();
```

File: lib/mshoplib/src/MShop/Plugin/Provider/Order/PropertyAdd.php [hunk 3]
```diff
@@ -181,15 +194,15 @@ protected function addAttributes( \Aimeos\MShop\Order\Item\Base\Product\Iface $o
...
+	protected function getProductItems( array $productCodes ) : \Aimeos\Map
 	{
 		$manager = \Aimeos\MShop::create( $this->getContext(), 'product' );
 		$search = $manager->filter( true );
 		$expr = [
-			$search->compare( '==', 'product.id', array_unique( $productIds ) ),
+			$search->compare( '==', 'product.code', array_unique( $productCodes ) ),
 			$search->getConditions(),
 		];
 		$search->setConditions( $search->and( $expr ) );
```
Comment (aimeos) [inline lib/mshoplib/src/MShop/Plugin/Provider/Order/PropertyAdd.php:208]: You don't use both, IDs and codes

File: lib/mshoplib/src/MShop/Plugin/Provider/Order/PropertyAdd.php [hunk 4]
```diff
@@ -157,9 +158,21 @@ protected function addAttributes( \Aimeos\MShop\Order\Item\Base\Product\Iface $o
...
 		}
 
+		if($products->count() > 1) {
+			$variant = $products->find(
+				function ( \Aimeos\MShop\Product\Item\Iface $item ) use ( $orderProduct ) : bool {
+					return $item->getCode() === $orderProduct->getProductCode();
+				}
+			);
+		}
+
```
Comment (aimeos) [inline lib/mshoplib/src/MShop/Plugin/Provider/Order/PropertyAdd.php:168]: It's more efficient if you use `$map = $products->col( null, 'product.code' );` first and then access the products by code

Comment (iammart) [reply lib/mshoplib/src/MShop/Plugin/Provider/Order/PropertyAdd.php:168]: Nice! Im used to using laravel collections and find myself digging through your docs for functions that are similar that I can use.

File: lib/mshoplib/src/MShop/Plugin/Provider/Order/PropertyAdd.php (hunk 3 above)
Comment (iammart) [reply lib/mshoplib/src/MShop/Plugin/Provider/Order/PropertyAdd.php:208]: My bad, I never copied my changes over, im copying my edits from editor to the browser.

Im trying to find more info on how I can run the test suite using phing, I just setup the environment so it has some data to work against.

File: lib/mshoplib/src/MShop/Plugin/Provider/Order/PropertyAdd.php (hunk 4 above)
Comment (aimeos) [reply lib/mshoplib/src/MShop/Plugin/Provider/Order/PropertyAdd.php:168]: The Map methods are very close to Laravel, only if standard PHP uses another name (`array_column()` instead of Laravel's `pluck()`), then the Map methods follow the PHP naming.

File: lib/mshoplib/src/MShop/Plugin/Provider/Order/PropertyAdd.php (hunk 2 above)
Comment (iammart) [reply lib/mshoplib/src/MShop/Plugin/Provider/Order/PropertyAdd.php:134]: Do we still need to keep the checks for orderProduct(s)? ie. can do you want me to also remove the loop?

Comment (aimeos) [reply lib/mshoplib/src/MShop/Plugin/Provider/Order/PropertyAdd.php:134]: Don't think they are really necessary

Comment (iammart) [reply lib/mshoplib/src/MShop/Plugin/Provider/Order/PropertyAdd.php:134]: > You can shorten this to:
> 
> ```
> $ids = map( $value )->getProductId();
> $codes = map( $value )->getProductCode();
> ```

This is sorcery...it's pretty cool to be able to call a method on an array of objects directly from the list.

Comment (aimeos) [reply lib/mshoplib/src/MShop/Plugin/Provider/Order/PropertyAdd.php:134]: jQuery style ;-)

File: lib/mshoplib/src/MShop/Plugin/Provider/Order/PropertyAdd.php [hunk 5]
```diff
@@ -121,18 +121,13 @@ public function update( \Aimeos\MW\Observer\Publisher\Iface $order, string $acti
 		if( !is_array( $value ) )
 		{
 			\Aimeos\MW\Common\Base::checkClass( \Aimeos\MShop\Order\Item\Base\Product\Iface::class, $value );
-			return $this->addAttributes( $value, $this->getProductItems( [$value->getProductId()] ), $types );
+			return $this->addAttributes( $value, $this->getProductItems( [$value->getProductId()], [$value->getProductCode()]), $types );
 		}
 
-		$list = [];
+		$ids = map( $value )->getProductId()->toArray();
+		$codes = map( $value )->getProductCode()->toArray();
```
Comment (aimeos) [inline lib/mshoplib/src/MShop/Plugin/Provider/Order/PropertyAdd.php:128]: You can use `->unique()` already here and pass the Map objects if you change the method signature of getProductsItems() to `iterable` instead of `array`

Comment (iammart) [reply lib/mshoplib/src/MShop/Plugin/Provider/Order/PropertyAdd.php]: I did do that initially, but my mistake was type hinting as Aimeos\Map, which would have failed for the call we make on the `addAttributes` method outside of the loop.

File: lib/mshoplib/src/MShop/Plugin/Provider/Order/PropertyAdd.php [hunk 6]
```diff
@@ -181,18 +185,19 @@ protected function addAttributes( \Aimeos\MShop\Order\Item\Base\Product\Iface $o
...
-			$search->getConditions(),
-		];
-		$search->setConditions( $search->and( $expr ) );
+
+		$search->add( $search->or( [
+			$search->is( 'product.id', '==', array_unique( $productIds ) ),
+			$search->is( 'product.code', '==', array_unique( $productCodes ) ),
+		] ) );
 
 		return $manager->search( $search, ['product/property'] );
```
Comment (aimeos) [inline lib/mshoplib/src/MShop/Plugin/Provider/Order/PropertyAdd.php:202]: We could save a lot of `$products->col( null, 'product.code' )` mappings in `addAttributes()` if we already return the merged properties already here. For that, we would need an array of product.code/product.id pairs, search for both like now and return a product.id/propertiy items map. What do you think?

Comment (iammart) [reply lib/mshoplib/src/MShop/Plugin/Provider/Order/PropertyAdd.php:202]: The method should probably change to reflect what it returns then.

`getProductItems` should probably become `getProductProperties`.

I guess we would also want to filter the attributes by the types configured here also, which means passing the types to method.

Do you not think passing an array of code/ids into `getProductProperties` is less explicit than the current approach, as the filter will need to extract the keys and values for the relevant columns to filter on?

Comment (aimeos) [reply lib/mshoplib/src/MShop/Plugin/Provider/Order/PropertyAdd.php:202]: We can also pass IDs and codes separately and return a map of product.code as keys and property items as values.
Your are right, we should rename the method, `getProperties()` is shorter and would fit too

Comment (iammart) [reply lib/mshoplib/src/MShop/Plugin/Provider/Order/PropertyAdd.php:202]: Do we need to return the map keyed as you described?

The changes ive made locally within the now labelled `getProperties` function, are to accept the types.

I map over the types and return a new map that includes the properties.

It returns something like the following:

```php
Aimeos\Map {#1780
  #list: array:3 [
    "package-weight" => Aimeos\Map {#1778
      #list: array:1 [
        182 => "0.4"
      ]
    }
    "package-net-weight" => Aimeos\Map {#1777
      #list: array:1 [
        190 => "0.4"
      ]
    }
    "package-dangerous" => Aimeos\Map {#1772
      #list: array:2 [
        191 => "0"
        192 => "0"
      ]
    }
  ]
}
```

This is then passed to `addAttributes` replacing the existing `$products` parameter, and I iterate over this and attach to the `$orderProduct`.

The tests are still passing for me with these changes - but could be a false positive due to coverage. I can publish my changes If you like?

Comment (aimeos) [reply lib/mshoplib/src/MShop/Plugin/Provider/Order/PropertyAdd.php:202]: Yes, that should work

File: lib/mshoplib/src/MShop/Plugin/Provider/Order/PropertyAdd.php [hunk 7]
```diff
@@ -141,55 +141,35 @@ public function update( \Aimeos\MW\Observer\Publisher\Iface $order, string $acti
...
-
-		foreach( $types as $type )
-		{
-			$list = $product->getProperties( $type );
+		$properties->each(function ( $attributes, $type ) use ( &$orderProduct ) {
 
-			if( $variant ?? false ) {
-				$list->union( $variant->getProperties( $type ) );
+			if ( ($attrItem = $orderProduct->getAttributeItem($type, 'product/property')) === null ) {
+				$attrItem = $this->orderAttrManager->create();
```
Comment (aimeos) [inline lib/mshoplib/src/MShop/Plugin/Provider/Order/PropertyAdd.php:153]: You can shorten this to:
```
$attrItem = $orderProduct->getAttributeItem( $type, 'product/property' ) ?: $this->orderAttrManager->create();
```

File: lib/mshoplib/src/MShop/Plugin/Provider/Order/PropertyAdd.php [hunk 8]
```diff
@@ -141,55 +141,35 @@ public function update( \Aimeos\MW\Observer\Publisher\Iface $order, string $acti
...
-
-		if( $products->count() > 1 ) {
-			$variant = $products->col( null, 'product.code' )
-				->get( $orderProduct->getProductCode() );
-		}
-
-		foreach( $types as $type )
-		{
-			$list = $product->getProperties( $type );
+		$properties->each(function ( $attributes, $type ) use ( &$orderProduct ) {
```
Comment (aimeos) [inline lib/mshoplib/src/MShop/Plugin/Provider/Order/PropertyAdd.php:150]: Where is `$type` comming from? `&` isn't necessary for objects

File: lib/mshoplib/src/MShop/Plugin/Provider/Order/PropertyAdd.php [hunk 9]
```diff
@@ -199,6 +179,12 @@ protected function getProductItems( iterable $productIds, iterable $productCodes
...
 		] ) );
 
-		return $manager->search( $search, ['product/property'] );
+		$items = $manager->search( $search, ['product/property'] );
+
+		return map( $types )->map( function ( $type ) use ( $items ) {
+			if( !( $properties = $items->getProperties($type)->collapse() )->empty() ) {
+				return [ $type => $properties ];
+			}
+		} )->filter()->collapse(1);
```
Comment (aimeos) [inline lib/mshoplib/src/MShop/Plugin/Provider/Order/PropertyAdd.php:188]: This should be sufficient to get a product code index and the properties as list of items:
```
return $manager->search( $search, ['product/property'] )->col( null, 'product.code' )->getProperties( $types );
```

File: lib/mshoplib/src/MShop/Plugin/Provider/Order/PropertyAdd.php (hunk 8 above)
Comment (iammart) [reply lib/mshoplib/src/MShop/Plugin/Provider/Order/PropertyAdd.php:150]: `$type` is the key in the properties map, it's the type the user configures in the admin for the plugin.

An example of the properties passed into `addAttributes` is below.

```php
Array
(
    [package-width] => Aimeos\Map Object
        (
            [list:protected] => Array
                (
                    [7] => 15.0
                )

            [sep:protected] => /
        )

)
```

File: lib/mshoplib/src/MShop/Plugin/Provider/Order/PropertyAdd.php (hunk 9 above)
Comment (iammart) [reply lib/mshoplib/src/MShop/Plugin/Provider/Order/PropertyAdd.php:188]: `getProperties` requires a string not an array, does it accept comma separated values? If so we would need to implode `$types`.

**Comment by iammart** [comment]
How about the following?

```php

$manager->search( $search, ['product/property'] )
    ->col( null, 'product.code' )
    ->map( function( $product ) use( $types ) {
        foreach ( $types as $type ) {
            return $product->getProperties( $type );
        }
    });
```

**Comment by aimeos** [comment]
> $manager->search( $search, ['product/property'] )
    ->col( null, 'product.code' )
    ->map( function( $product ) use( $types ) {
        foreach ( $types as $type ) {
            return $product->getProperties( $type );
        }
    });

That should work.

**Comment by aimeos** [comment]
@iammart Any news on this topic?
